x-pack/test/api_integration/apis/code/feature_controls.ts

```diff
@@ -4,67 +4,229 @@
+        endpoint.expectResponse(result);
+      }
+    }
+  }
 
   describe('feature controls', () => {
-    // TODO implement tests similar to APM
+    it(`APIs can't be accessed by .code-* read privileges role`, async () => {
+      const username = 'logstash_read';
+      const roleName = 'logstash_read';
+      const password = `${username}-password`;
+      try {
+        await security.role.create(roleName, {
+          elasticsearch: {
+            indices: [
+              {
+                names: ['.code-*'],
+                privileges: ['read'],
+              },
+            ],
+          },
+        });
+
+        await security.user.create(username, {
+          password,
+          roles: [roleName],
+          full_name: 'a kibana user',
+        });
+
+        await executeRequests(username, password, 'forbidden');
+      } finally {
+        await security.role.delete(roleName);
+        await security.user.delete(username);
+      }
+    });
+
+    it(`APIs can be accessed global all with .code-* read privileges role`, async () => {
+      const username = 'global_all';
+      const roleName = 'global_all';
+      const password = `${username}-password`;
+      try {
+        await security.role.create(roleName, {
+          elasticsearch: {
+            indices: [
+              {
+                names: ['.code-*'],
+                privileges: ['read'],
+              },
+            ],
+          },
+          kibana: [
+            {
+              base: ['all'],
+              spaces: ['*'],
+            },
+          ],
+        });
+
+        await security.user.create(username, {
+          password,
+          roles: [roleName],
+          full_name: 'a kibana user',
+        });
+
+        await executeRequests(username, password, 'response');
+      } finally {
+        await security.role.delete(roleName);
+        await security.user.delete(username);
+      }
+    });
+
+    it(`APIs can't be accessed by dashboard all with .code-* read privileges role`, async () => {
+      const username = 'dashboard_all';
+      const roleName = 'dashboard_all';
+      const password = `${username}-password`;
+      try {
+        await security.role.create(roleName, {
+          elasticsearch: {
+            indices: [
+              {
+                names: ['.code-*'],
+                privileges: ['read'],
+              },
+            ],
+          },
+          kibana: [
+            {
+              feature: {
+                dashboard: ['all'],
+              },
+              spaces: ['*'],
+            },
+          ],
+        });
+
+        await security.user.create(username, {
+          password,
+          roles: [roleName],
+          full_name: 'a kibana user',
+        });
+
+        await executeRequests(username, password, 'forbidden');
+      } finally {
+        await security.role.delete(roleName);
+        await security.user.delete(username);
+      }
+    });
+
+    // describe('spaces', () => {
```

Comment:
@kobelb @legrego since Code won't support space in the first version. Do we need to have tests for spaces?

Reply:
I think it's still worthwhile to have tests for spaces. Even though Code does not support spaces yet, we should still have tests which verify that Code behaves correctly when the user is not in the default space.

Reply:
done